mitmproxy/proxy/layers/quic.py

```diff
@@ -0,0 +1,1052 @@
+    flow: udp.UDPFlow | None  # used for datagrams and to signal general connection issues
+    quic_client: QuicConnection | None = None
+    quic_server: QuicConnection | None = None
+    streams_by_flow: dict[tcp.TCPFlow, QuicStream]
+    streams_by_id: dict[int, QuicStream]
+
+    def __init__(self, context: context.Context, ignore: bool = False) -> None:
+        super().__init__(context)
+        self.buffer_from_client = []
+        self.buffer_from_server = []
+        if ignore:
+            self.flow = None
+        else:
+            self.flow = udp.UDPFlow(self.context.client, self.context.server, live=True)
+        self.streams_by_flow = {}
+        self.streams_by_id = {}
+
+    def get_or_create_stream(
+        self, stream_id: int
+    ) -> layer.CommandGenerator[QuicStream]:
+        if stream_id in self.streams_by_id:
+            return self.streams_by_id[stream_id]
+        else:
+            # register the stream and start the flow
+            stream = QuicStream(self.context, stream_id, ignore=self.flow is None)
+            self.streams_by_id[stream.stream_id] = stream
+            if stream.flow is not None:
+                self.streams_by_flow[stream.flow] = stream
+                yield TcpStartHook(stream.flow)
+            return stream
+
+    def handle_quic_event(
+        self,
+        event: quic_events.QuicEvent,
+        from_client: bool,
+    ) -> layer.CommandGenerator[None]:
+        # buffer events if the peer is not ready yet
+        peer_quic = self.quic_server if from_client else self.quic_client
+        if peer_quic is None:
+            (self.buffer_from_client if from_client else self.buffer_from_server).append(event)
+            return
+        peer_connection = self.context.server if from_client else self.context.client
+
+        if isinstance(event, quic_events.DatagramFrameReceived):
+            # forward datagrams (that are not stream-bound)
+            if self.flow is not None:
+                udp_message = udp.UDPMessage(from_client, event.data)
+                self.flow.messages.append(udp_message)
+                yield UdpMessageHook(self.flow)
+                data = udp_message.content
+            else:
+                data = event.data
+            peer_quic.send_datagram_frame(data)
+
+        elif isinstance(event, quic_events.StreamDataReceived):
+            # ignore data received from already ended streams
+            stream = yield from self.get_or_create_stream(event.stream_id)
+            if stream.has_ended(from_client):
+                yield commands.Log(f"Received {len(event.data)} byte(s) on already closed stream #{event.stream_id}.", level="debug")
+                return
+
+            # forward the message allowing addons to change it
+            if stream.flow is not None:
+                tcp_message = tcp.TCPMessage(from_client, event.data)
+                stream.flow.messages.append(tcp_message)
+                yield TcpMessageHook(stream.flow)
```

Comment:
FYI one reason why HttpStream subclasses layer.Layer is that streams can block independently. Here we handle all message hooks in QuicStreamLayer (the HTTP equivalent would be HttpLayer), which means that one intercepted stream blocks all other streams.

This here is definitely good enough for a first iteration though.

Reply:
Well in that case... I was actually exploring the idea of doing `QuicStreamLayer` differently before, namely splitting up the  client connection per stream, and merging it back together at the server. That would allow for doing `NextLayer` per stream (implementing DNS over QUIC would be a breeze), but also ease implementation of CONNECT/upstream support. With the new multi-server mode I dropped it, because of how connections are registered now, since I planned to use proxyserver's connections for routing. But if the additional complexity of layer routing like in HTTP is desired anyways, I'll give it another try :)